File: app/src/main/java/se/insektionen/songbook/ui/SongbookListAdapter.java
package se.insektionen.songbook.ui;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import se.insektionen.songbook.R;
import se.insektionen.songbook.model.Song;

/**
 * List adapter for the list of songs in a songbook.
 */
public final class SongbookListAdapter extends BaseAdapter implements Filterable {
	public static final String CATEGORY_QUERY = "?category=";
	private static final String ELLIPSIS = "...";
	private static final String[] TRIM_END = new String[]{".", ",", "!", "?"};
	private static final String[] TRIM_START = new String[]{"*", "//:"};
	private final LayoutInflater mInflater;
	private final List<Song> mList;
	private final Filter mFilter = new SongbookListFilter();
	private List<Song> mFilteredList;

	public SongbookListAdapter(Context context, List<Song> list) {
		mList = list;
		mFilteredList = list;
		mInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
	}

	@Override
	public int getCount() {
		return mFilteredList.size();
	}

	@Override
	public Filter getFilter() {
		return mFilter;
	}

	@Override
	public Object getItem(int position) {
		return mFilteredList.get(position);
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent) {
		View view;
		if (null == convertView) {
			view = mInflater.inflate(R.layout.list_item_song, parent, false);
		} else {
			view = convertView;
		}

		Song song = mFilteredList.get(position);

		TextView nameText = (TextView) view.findViewById(R.id.song_list_primary);
		nameText.setText(song.getName());

		TextView firstLineText = (TextView) view.findViewById(R.id.song_list_secondary);
		firstLineText.setText(formatFirstLineOfSong(song.firstLineOfSong()));

		TextView categoryText = (TextView) view.findViewById(R.id.song_list_tertiary);
		categoryText.setText(song.getCategory());

		return view;
	}

	private String formatFirstLineOfSong(String line) {
		for (String trimStart : TRIM_START) {
			if (line.startsWith(trimStart)) {
				line = line.substring(trimStart.length());
			}
		}
		for (String trimEnd : TRIM_END) {
			if (line.endsWith(trimEnd)) {
				line = line.substring(0, line.length() - trimEnd.length());
			}
		}

		return line.trim().concat(ELLIPSIS);
	}

	private final class SongbookListFilter extends Filter {
		@Override
		protected FilterResults performFiltering(CharSequence constraint) {
			FilterResults results = new FilterResults();
			List<Song> filteredList;

			if (!TextUtils.isEmpty(constraint)) {
				String categoryFilter = null;
				String textFilter;

				/*
				 * The query is of the format {textFilter}?category={categoryFilter}
				 * Both parts are optional.
				 */

				String query = constraint.toString();
				if (query.contains(CATEGORY_QUERY)) {
					int idx = query.indexOf(CATEGORY_QUERY);
					categoryFilter = query.substring(idx + CATEGORY_QUERY.length());
					textFilter = query.substring(0, idx);
				} else {
					textFilter = query;
				}

				filteredList = new ArrayList<>();
				for (Song s : mList) {
					if (filter(s, textFilter, categoryFilter)) {
						filteredList.add(s);
					}
				}
			} else {
				filteredList = mList;
			}

			results.values = filteredList;
			results.count = filteredList.size();
			return results;
		}

		@Override
		protected void publishResults(CharSequence constraint, FilterResults results) {
			//noinspection unchecked
			mFilteredList = (List<Song>) results.values;
			if (results.count > 0) {
				notifyDataSetChanged();
			} else {
				notifyDataSetInvalidated();
			}
		}

		private boolean filter(Song song, String textFilter, String categoryFilter) {
			if (!TextUtils.isEmpty(textFilter) && !song.matches(textFilter.toLowerCase())) {
				return false;
			}
			return TextUtils.isEmpty(categoryFilter) || song.getCategory().equals(categoryFilter);
		}
	}
}

File: app/src/main/java/se/insektionen/songbook/ui/SongbookListAdapter.kt
package se.insektionen.songbook.ui

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Filter
import android.widget.Filterable
import android.widget.TextView
import androidx.core.content.getSystemService
import se.insektionen.songbook.R
import se.insektionen.songbook.model.Song
import java.util.Locale

/**
 * List adapter for the list of songs in a songbook.
 */
class SongbookListAdapter(
    context: Context,
    initialList: List<Song>
) : BaseAdapter(), Filterable {
    private val filter: Filter = SongbookListFilter()
    private var filteredList: List<Song> = initialList
    private val inflater: LayoutInflater = context.getSystemService()!!
    private val list: List<Song> = initialList

    override fun getCount(): Int = filteredList.size

    override fun getFilter(): Filter = filter

    override fun getItem(position: Int): Any = filteredList[position]

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        val view: View = convertView ?: inflater.inflate(R.layout.list_item_song, parent, false)
        val song = filteredList[position]
        view.findViewById<TextView>(R.id.song_list_primary).apply { text = song.name }
        view.findViewById<TextView>(R.id.song_list_secondary)
            .apply { text = formatFirstLineOfSong(song.firstLineOfSong()) }
        view.findViewById<TextView>(R.id.song_list_tertiary).apply { text = song.category }
        return view
    }

    private fun formatFirstLineOfSong(line: String): String {
        var result = line
        TRIM_START.forEach {
            if (result.startsWith(it)) {
                result = result.substring(it.length)
            }
        }
        TRIM_END.forEach {
            if (result.endsWith(it)) {
                result = result.substring(0, result.length - it.length)
            }
        }
        return result.trim() + ELLIPSIS
    }

    private inner class SongbookListFilter : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val resultList =
                if (constraint.isNotEmpty()) {
                    /*
                     * The query is of the format {textFilter}?category={categoryFilter}
                     * Both parts are optional.
                     */
                    val query = constraint.toString()
                    val categoryFilter: String
                    val textFilter: String

                    if (query.contains(CATEGORY_QUERY)) {
                        val idx = query.indexOf(CATEGORY_QUERY)
                        categoryFilter = query.substring(idx + CATEGORY_QUERY.length)
                        textFilter = query.substring(0, idx)
                    } else {
                        categoryFilter = ""
                        textFilter = query
                    }
                    list.filter { filter(it, textFilter, categoryFilter) }
                } else {
                    list
                }

            return FilterResults().also {
                it.values = resultList
                it.count = resultList.size
            }
        }

        @Suppress("UNCHECKED_CAST")
        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            filteredList = results.values as List<Song>
            if (results.count > 0) {
                notifyDataSetChanged()
            } else {
                notifyDataSetInvalidated()
            }
        }

        private fun filter(song: Song, textFilter: String, categoryFilter: String): Boolean {
            if (categoryFilter.isNotEmpty() && song.category != categoryFilter) {
                // Drop any song that does not match the selected category
                return false
            }
            if (textFilter.isEmpty()) {
                // Accept any song if the text filter is empty
                return true
            }
            // Otherwise, apply the text filter
            return textFilter.isNotEmpty() && song.matches(textFilter.lowercase(Locale.getDefault()))
        }
    }

    companion object {
        const val CATEGORY_QUERY = "?category="
        private const val ELLIPSIS = "..."
        private val TRIM_END = arrayOf(".", ",", "!", "?")
        private val TRIM_START = arrayOf("*", "//:")
    }
}
